**Step 1: Import packages**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

**Step 2: Define working directories**

In [3]:
os.chdir('C:/Users/user/Documents/Berkeley_MIDS/W207_Machine_Learning/w207_final_project')

**Step 3: Define classes**

**Step 4: Define functions**

**Step 5: Read data**

In [6]:
# Load npz files to combine and print content files
data_1 = np.load('et_w207_project_npz_files_5000_tmp_tmp5ocrhjnn.npz')
data_1.close

data_2 = np.load('et_w207_project_npz_files_5000_tmp_tmp092sag67.npz')
data_2.close

data_3 = np.load('et_w207_project_npz_files_5000_tmp_tmpf31_pn8p.npz')
data_3.close

data_4 = np.load('et_w207_project_npz_files_5000_tmp_tmpq5b2g4n2.npz')
data_4.close

print(data_1.files)

['arr_0', 'arr_1']


In [7]:
# Show file arr_0 example
data_1['arr_0'][:5]

array(['ozbpir-9ey6js-ggdqwo', '378ykanq', 'p366md-big834-7o23k6',
       '3zruwvl2', 'q18iae-3vnh74-79npmy'], dtype='<U20')

In [8]:
# Show file arr_1 example
data_1['arr_1'][:5]

array([[-103.939   , -116.779   , -123.68    , ..., -103.939   ,
        -116.779   , -123.68    ],
       [ -46.939003,  -74.779   , -118.68    , ...,  -57.939003,
         -84.779   , -118.68    ],
       [ 151.061   ,  138.22101 ,  131.32    , ...,  151.061   ,
         138.22101 ,  131.32    ],
       [  -3.939003,  -66.779   ,  -96.68    , ...,   -3.939003,
         -66.779   ,  -96.68    ],
       [-103.939   , -116.779   , -123.68    , ..., -103.939   ,
        -116.779   , -123.68    ]], dtype=float32)

In [9]:
# Get number of observations in each file
print(data_1['arr_0'].shape)
print(data_2['arr_0'].shape)
print(data_3['arr_0'].shape)
print(data_4['arr_0'].shape)

(878,)
(893,)
(638,)
(1353,)


In [10]:
# Concatenate data from all files
data_key = np.concatenate((data_1['arr_0'], data_2['arr_0'], data_3['arr_0'], data_4['arr_0']))

data_pixel = np.concatenate((data_1['arr_1'], data_2['arr_1'], data_3['arr_1'], data_4['arr_1']))
print(data_pixel.shape)

(3762, 150528)


In [15]:
# Store pixel vectors as DF with attachment key

# Create key value pairs with attachment key and pixel array
data_dict = {}
for i in range(data_key.shape[0]):
    data_dict[data_key[i]] = [data_pixel[i]]

df_pixel = pd.DataFrame(data=data_dict)
df_pixel_t = df_pixel.transpose()
df_pixel_t.index.names = ['attachment_key']
df_pixel_t.columns = ['pixel_array']
df_pixel_t

pixel_array
attachment_key                                                         
ozbpir-9ey6js-ggdqwo  [-103.939, -116.779, -123.68, -103.939, -116.7...
378ykanq              [-46.939003, -74.779, -118.68, -24.939003, -57...
p366md-big834-7o23k6  [151.061, 138.22101, 131.32, 151.061, 138.2210...
3zruwvl2              [-3.939003, -66.779, -96.68, -3.939003, -66.77...
q18iae-3vnh74-79npmy  [-103.939, -116.779, -123.68, -103.939, -116.7...
...                                                                 ...
pktyig-481wyg-6gyulf  [-91.939, -104.779, -97.68, -92.939, -105.779,...
zu7y3rha              [-1.939003, -14.778999, -21.68, -1.939003, -14...
pxcwxf-8tnpyg-8vhb9r  [77.061, 46.221, 32.32, 79.061, 49.221, 33.32,...
p4hakf-fsskmw-1224up  [-102.939, -115.779, -122.68, -98.939, -111.77...
ptuhsk-22124o-f0sxoc  [112.061, 100.221, 89.32, 114.061, 102.221, 93...

[3759 rows x 1 columns]

In [16]:
# Load excel data
df = pd.read_csv('school_project_data_set-query.csv')
df.head()

attachment_key    size_mb  height   width    h_to_w  \
0  q36z0t-gf0ta0-91uq4d   5.256976  3351.0  2530.0  1.324506   
1  p0td2i-8fv5ew-2c7suq   0.688739  3420.0  1365.0  2.505495   
2  q6p0vh-6wed5k-3c8m2s  11.636952  6240.0  4160.0  1.500000   
3  pg1c3o-fqyk3c-72hr82   4.161787  3257.0  4885.0  0.666735   
4  q52mox-a2mcsg-4ir06w   4.081567  3738.0  5600.0  0.667500   

                            filename  logo  
0               WVC021819TH-5756.jpg     0  
1  Lastricato_CastelloDelTrebbio.jpg     0  
2              WVC2Port2020-9684.jpg     0  
3                       EB4A1971.jpg     0  
4                      934879096.jpg     0

In [17]:
# Merge dataframes by attachment key
merge_meta_pixel = df.merge(df_pixel_t, on='attachment_key')
merge_meta_pixel

attachment_key   size_mb  height   width    h_to_w  \
0     q4zhny-2mxzu8-fz6s49  2.423621  4500.0  4500.0  1.000000   
1     pgeoak-9l272o-5dah6f  0.014292   137.0   327.0  0.418960   
2     pgeoak-9l272o-5dah6f  0.014292   137.0   327.0  0.418960   
3                 vd92yz5a  0.109244   716.0  1428.0  0.501401   
4     pu19n0-ajc888-actmf5  0.027520  1251.0  1250.0  1.000800   
...                    ...       ...     ...     ...       ...   
4422   px1yxs-mfwvk-az49zm  0.045961   333.0   500.0  0.666000   
4423  q6mxyb-ferzig-bai2df  0.220249   352.0   500.0  0.704000   
4424  olffz6-6619ps-53fw6l  0.006069   162.0   500.0  0.324000   
4425   px1yxs-mfwvk-174b2s  0.071568   540.0   500.0  1.080000   
4426  q6qjrl-fs5ej4-c6p902  0.073736   322.0   500.0  0.644000   

                                      filename  logo  \
0                    Ska-HazyIPA-can-White.jpg     0   
1     Screen Shot 2018-10-10 at 7.07.12 PM.png     0   
2     Screen Shot 2018-10-10 at 7.07.12 PM.png     1   
3                               a pha_logo.jpg     1   
4        AXON - Icon Family - July 2019-03.png     1   
...                                        ...   ...   
4422           zoom_se_rts400_574634_a_15a.jpg     0   
4423    zoom_se_etsec150_576329_sod_01b_US.jpg     0   
4424           web-logo-run-gum-energy-gum.png     1   
4425              zoom_fr_sys_498889_z_01a.jpg     0   
4426         zoom_se_dts400_576064_sod_01a.jpg     0   

                                            pixel_array  
0     [151.061, 138.22101, 131.32, 151.061, 138.2210...  
1     [140.061, 133.22101, 125.32, 140.061, 133.2210...  
2     [140.061, 133.22101, 125.32, 140.061, 133.2210...  
3     [151.061, 138.22101, 131.32, 151.061, 138.2210...  
4     [-103.939, -116.779, -123.68, -103.939, -116.7...  
...                                                 ...  
4422  [94.061, 94.221, 101.32, 94.061, 94.221, 101.3...  
4423  [151.061, 138.22101, 131.32, 151.061, 138.2210...  
4424  [-103.939, -116.779, -123.68, -103.939, -116.7...  
4425  [151.061, 138.22101, 131.32, 151.061, 138.2210...  
4426  [151.061, 138.22101, 131.32, 151.061, 138.2210...  

[4427 rows x 8 columns]

In [ ]:
# STILL WORKING
# Split into train, development, test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4,
                                                    random_state=1, stratify=y)

**Step 6: Data preprocessing**

In [ ]:
# Standardize pixel values between 0 and 1 by dividing by 255

**Step 7: Single Layer NN**

7.1: Build the NN Model

7.2: Fit the NN Model

7.3: Evaluate performance on the test set

7.4: Make predictions

**Step 8: Multilayer NN**

8.1: Build the NN Model

8.2: Fit the NN Model

8.3: Evaluate performance on the test set

8.4: Make predictions

**Conclusions**